File: rust/src/include/math.rs
use std::cmp::PartialOrd;
use std::fmt::Debug;
use std::mem::size_of;

use num_traits::NumAssign;
use num_traits::one;
use num_traits::CheckedMul;

const MAX_FACTORIAL: [u8; 16] = [
    5, // u8
    8, // u16
    10, 12, // u32
    14, 16, 18, 20, // u64
    22, 24, 25, 27, 29, 30, 32, 34 // u128
];

pub fn factorial<I>(n: u8) -> I
where I: NumAssign + From<u8>
{
    let mut answer: I = one();
    for i in 2..=n {
        answer *= i.into();
    }
    return answer
}

pub fn n_choose_r<I>(n: usize, r: usize) -> I
where I: Copy + From<u8> + From<u64> + NumAssign + PartialOrd + CheckedMul + Debug
{
    if n < r {
        panic!("Out of function's bounds");
    }
    if n < MAX_FACTORIAL[size_of::<I>() - 1] as usize {
        return factorial::<I>(n as u8) / factorial::<I>(r as u8) / factorial::<I>((n - r) as u8);
    }
    // slow path for larger numbers
    let mut answer: I = one();
    let mut factors: Vec<i8> = vec![1; n + 1];
    factors[0] = 0;
    factors[1] = 0;
    // negative factor values indicate need to divide
    for i in 2..=r {
        factors[i] -= 1;
    }
    for i in 2..=(n - r) {
        factors[i] -= 1;
    }
    // this loop reduces to prime factors only
    for i in (1..n).rev() {
        for j in 2..i {
            if i % j == 0 {
                factors[j] += factors[i];
                factors[i / j] += factors[i];
                factors[i] = 0;
                break;
            }
        }
    }
    let mut i: usize = 2;
    let mut j: usize = 2;
    while i <= n {
        while factors[i] > 0 {
            let mut result = answer.checked_mul(&(i as u64).into());
            while result.is_none() && j <= n {
                while factors[j] < 0 {
                    answer /= (j as u64).into();
                    factors[j] += 1;
                }
                j += 1;
                result = answer.checked_mul(&(i as u64).into());
            }
            factors[i] -= 1;
            answer = result.unwrap_or_else(|| panic!("nCr overflow: {:?} * {}", answer, i));
        }
        i += 1;
    }
    while j <= n {
        while factors[j] < 0 {
            answer /= (j as u64).into();
            factors[j] += 1;
        }
        j += 1;
    }
    return answer;
}

File: fortran/src/include/math.f90
module math
    use constants
    implicit none
contains
    pure integer(i18t) function factorial(n) result(answer)
        integer, intent(in) :: n
        integer :: i
        answer = 1
        do i = 2, n
            answer = answer * i
        end do
    end function

    pure integer(i18t) function n_choose_r(n, r) result(answer)
        integer, intent(in) :: n, r
        if (n <= MAX_FACTORIAL_64) then
            answer = factorial(n) / factorial(r) / factorial(n-r)  ! fast path if small enough
        else
            answer = n_choose_r_slow(n, r)  ! slow path for larger numbers
        end if
    end function

    integer(i18t) function n_choose_r_slow(n, r) result(answer)
        integer, intent(in) :: n, r
        integer(i2t), allocatable :: factors(:)
        integer(i18t) :: tmp
        integer :: i, j
        allocate(factors(n))
        if (.not. allocated(factors)) then
            print *, "n_choose_r allocation failed. Exiting."
            stop ERROR_ALLOCATE_FAILED
        end if
        factors = 0
        ! collect factors of final number
        do i = 2, n
            factors(i) = 1
        end do
        ! negative factor values indicate need to divide
        do i = 2, r
            factors(i) = factors(i) - 1_i2t
        end do
        do i = 2, n - r
            factors(i) = factors(i) - 1_i2t
        end do
        do i = n, 2, -1  ! this loop reduces to prime factors only
            do j = 2, i - 1
                if (mod(i, j) == 0) then
                    factors(j) += factors(i)
                    factors(i / j) += factors(i)
                    factors(i) = 0;
                    exit
                end if
            end do
        end do
        i = 2
        j = 2
        answer = 1
        do while (i <= n)
            do while (factors(i) > 0)
                tmp = answer
                answer = answer * i
                do while (answer < tmp .and. j <= n)
                    do while (factors(j) < 0)
                        tmp = tmp / j
                        factors(j) = factors(j) + 1_i2t
                    end do
                    j = j + 1_i2t
                    answer = tmp * i
                end do
                if (answer < tmp) then
                    answer = -1  ! this indicates an overflow
                    return
                end if
                factors(i) = factors(i) - 1_i2t
            end do
            i = i + 1_i2t
        end do
        do while (j <= n)
            do while (factors(j) < 0)
                answer = answer / j
                factors(j) = factors(j) + 1_i2t
            end do
            j = j + 1_i2t
        end do
        deallocate(factors)
    end function
end module math

